### ISSUE TITLE
AIR chiplet for arithmetic circuit evaluation 

### ISSUE THREAD

**Issue by Al-Kindi-0** [0xPolygonMiden/miden-vm#1610]
### Feature description

# Introduction

A large portion of the recursive verifier work goes into the constraint evaluation step. This usually entails evaluating a large algebraic expression over the extension field. If we think of this algebraic expression as an arithmetic circuit then this circuit has as inputs:
1. The out-of-domain (OOD) point $z$
2. The OOD trace columns evaluation frame: these are the (claimed) evaluations of the committed trace polynomials at the OOD point $z$ and $gz$
3. The OOD constraint columns evaluation frame: these are the (claimed) evaluations of the committed quotient segment polynomials at the OOD point $z$
4. The constraint composition randomness: these are the values used for combining the different constraints
5. The public inputs: these are values (over base field) and are usually inputs to boundary constraints
6. The divisors: these are expressions of the form $\frac{1}{z^a - b}$

The output of the circuit is then a single value (in the extension field) which should then be constrained to be equal to zero.

Hence, the problem of constraints evaluation can be described by the following components:

1. Inputs: these are the input values to the circuit
2. Arithmetic circuit description:
   1. Gate's type: we can assume that we have only addition and multiplication gates
   2. Circuit topology: each gate's inputs and outputs.

# An AIR for evaluating arithmetic circuits
Assume throughout that each gate is fan-in 2 but fan-out any positive integer.

Now, suppose that the description of the circuit is stored in memory as a sequence of tuples describing each of the gates where each tuple is made up of
1. `op`: either `add` or `mul` describing the type of the gate
2. `[l_inp_id, r_inp_id]`: the ids of the left and right input wires to the gate
3. `out_id`: the id of the current gate which also serves as the id of the output wires of the gate
4. `m`: a multiplicity parameter describing the fan-out of the gate
   
From a user point of view, the circuit will be specified by a digest and loaded to memory from the advice provider by un-hashing the said digest. Let `circuit_ptr` be a pointer to the memory region containing this circuit description.
As for the inputs to the circuit, we assume that they are stored in a contiguous region of memory, and let `input_ptr` be a pointer to this memory region.

### Layout

The following is a pictorial description of the layout of the chiplet. It can be thought of as being composed of three portions:

![eval_chiplet](https://github.com/user-attachments/assets/963c31b6-5049-46a1-abc9-f246dca8db2b)

1. Selectors: these are the selectors used to switch off between different modes of the chiplet. See below.
2. Hardcoded portion:
   1. `op`, `[l_inp_id, r_inp_id]`, `out_id` and `m` are as described above.
3. Dynamic portion:
   1. `[l_inp_val, r_inp_val]` are the current gates left and right input values. These are over the extension field over which the constraints evaluation is done.
   2. `out_val` is the current gates output values. These are over the extension field as well.
   3. `input_ptr` is the pointer to the inputs as described above.
   4. `circuit_ptr` is the pointer to the circuit description as described above.
   5. `clk` is the clock cycle at which the memory access' happen.
   6. `ctx` is the context at which the memory access' happen.

Note that there is an extra bus that will be used to wire the gates. The idea is for each gate to send `output` requests of the form `(out_id, out_val)` with multiplicity `m`. Similarly each gate sends a `input` request of the form `(l_inp_id, l_inp_val)` and `(r_inp_id, r_inp_val)`. We will use selectors to turn on/off these requests, see next section. Moreover, we will append a label to each type of interaction with the bus in order to distinguish them.

#### Loading inputs
This is indicated by a flag `f_load` derived from `selectors`. Here the chiplet makes read requests `(MEM_READ_OP_LABEL, clk, ctx, input_ptr)` to memory and populates `out_val` with the received values representing the circuit inputs.
The chiplet also makes read requests `(MEM_READ_OP_LABEL, clk, ctx, circuit_ptr)` to memory and populates `op`, `[l_inp_id, r_inp_id]`, `out_id` and `m`. In practice and at this step, the only relevant data is `m` as `op`, `[l_inp_id, r_inp_id]` are not relevant at this stage and `out_id` can be "computed" by the chiplet by starting from `0` and incrementing it by `1` for each subsequent row while `f_load == 1`.
At this stage we do not enforce any constraints but for the outputting of the gates values. This is done by sending `(out_id, out_val)` with multiplicity `m` to the wiring bus. 
Both `input_ptr` and `circuit_ptr` are incremented after each request.
When the `Loading` stage ends, which is determined when the next row is the first row of the next stage, the chiplet sends a message to the decoder with the current value of the `input_ptr` pointer, call it `input_end_ptr`, to indicate that it has finished loading the inputs.

#### Evaluating
This is indicated by a flag `f_eval` derived from `selectors` and mutually exclusive with `f_load`. Here the chiplet stops reading the inputs but continues reading the circuit description but now reads the full description of each gate i.e., `op`, `[l_inp_id, r_inp_id]`, `out_id` and `m`. Moreover, we now impose the gate constraint `out_val = l_inp_val * r_inp_val` or `out_val = l_inp_val + r_inp_val` depending on `op` where the addition/product is over the extension field.
From the wiring point of view, the chiplet makes, to the wiring bus, for each gate, requests to get inputs `(l_inp_id, l_inp_val)` and `(r_inp_id, r_inp_val)` and requests to output `(out_id, out_val)` with multiplicity `m`.

#### Outputting
This is indicated by a flag `f_out` derived from `selectors` and mutually exclusive with `f_eval` and `f_load`. Here the chiplet stops reading the inputs and circuit description. `f_out` also indicates that the `out_val` of the previous gate is the output of the circuit. Given the above, we can add a constraint for `out_val` to be zero or just output it by including it in the final message to the decoder. This message will include the final value of `circuit_ptr`, call it `circuit_end_ptr`, to signal that the chiplet has read and evaluated all of the circuit.

### Usage

We will assume that the inputs have already been loaded to memory which is the case for the applications we have in mind i.e., recursive verification. Then the program wishing to evaluate the circuit on the aforementioned inputs will un-hash the circuit to some region of memory. At this point the program calls the instruction that does the evaluation, say `eval_circuit`, with inputs `input_ptr`, `circuit_ptr`, `input_end_ptr` and `circuit_end_ptr` and this will correspond to the decoder making a request to the circuit evaluation chiplet of the form `(BEGIN_EVAL_OP_LABEL, clk, ctx, input_ptr)` and `(BEGIN_EVAL_OP_LABEL, clk, ctx, circuit_ptr)` to initialize the evaluation and `(END_EVAL_OP_LABEL, clk, ctx, input_end_ptr)` and `(END_EVAL_OP_LABEL, clk, ctx, circuit_end_ptr)` to request a response from the chiplet, where `*_end_ptr` is, as before, the value of the pointer at the boundary of the memory region of either the circuit inputs or the circuit description.
The chiplet will then respond first by initializing circuit evaluation and for this we will probably need a fourth flag, say `f_init`. Here, the chiplet will make responses to the decoder `(BEGIN_EVAL_OP_LABEL, clk, ctx, input_ptr)` and `(BEGIN_EVAL_OP_LABEL, clk, ctx, circuit_ptr)` to indicate the start of the circuit evaluation. The next chiplet row will then be the first row in the `Loading` phase.
After executing `Loading`, the chiplet will make the response  `(END_EVAL_OP_LABEL, clk, ctx, input_end_ptr)` to signal that it has completed loading the circuit inputs from memory.

Next comes the `Evaluating` phase after which the chiplet transitions to the `Outputting` phase and here it will issue a response `(END_EVAL_OP_LABEL, clk, ctx, circuit_end_ptr)` to the decoder indicating that it is done evaluating the circuit, which will imply that the circuit evaluated to zero or, if we choose to send the evaluation to the decoder, we can include it in this last message.

# Remarks and Open Questions:
1. A full description of the flags is TBD.
2. A description of the constraints on the flags is TBD. We are going to probably need constraints for correctly transitioning the circuit evaluation chiplet from one state to the other in a set sequential order.
3. It might be better to make the gates more expressive at the cost of adding more columns. The rational for this would be to decrease the gate count and hence the chiplet's execution trace length.

### Why is this feature needed?

As the VM's complexity increases, more AIR constraints are added which in turns translates to more work for the recursive verifier at the constraints evaluation step. Thus we need to find a way to accelerate this step and the current issue aims at investigating how to do so.

**Comment by bobbinth** [comment]
Looks great! Thank you for such a detailed description! A few thoughts/comments:

### Gate definition

One of the things we probably want to do is to make sure that we can encode a gate in a single field element. This way, we can increase the efficiency of "unhashing" of a circuit in memory. From your description, a gate is a tuple `(op, l_in_id, r_in_id, out_id, m)`. To fit into a filed element, we'd need to limit `l_in_id`, `r_in_id`, and `out_id` to something like $2^{16}$ and `m` to $2^{14}$.

Alternatively, we could skip $m$ (the prover will need to fill it in in the chiplet) and then `l_in_id`, `r_in_id`, and `out_id` could be limited by $2^{20}$.

### Selectors
We could avoid selectors by splitting loading inputs, evaluating, and reading output into 3 separate instructions (or maybe evaluating and reading output could be combined into a single instruction). The instructions could look like so:

```
#! Inputs: [input_ptr, input_end_ptr, eval_id]
#! Output: [input_ptr + 1, input_end_ptr, eval_id]
load_input

#! Inputs: [circuit_ptr, circuit_end_ptr, eval_id]
#! Output: [result_1, result_0, eval_id]
eval_circuit
```
Here, `eval_id` would be some unique value that the user would pick for a given evaluation (e.g., it could be based on the clock cycle).

The user would the load inputs into the evaluator bus by executing `load_inputs` in a loop, and then call `eval_circuit` using the same `eval_id`.

This way, the only memory reads the chiplet needs to do is to load the gates. This may simplify things a bit, though not sure by how much.

Another potential option is to encode all actions into `op`. This way, in addition to `add` and `mul` we'd also have `read` operation which would read an input from a given memory location.

### Input format
One thing that is not entirely clear to me is what's the best way to arrange inputs in memory. I believe currently we "interleave" current/next row values of the OOD point in a single word. Would this still work? Or would we need to make it so that each input value is stored in a separate word?

### Reading gates from memory
If I understood correctly, the described design assumes that we do 1 memory read of a single element per gate (i.e., per row of the chiplet). Since we can actually read 4 elements per memory read, I wonder if there is some way we can improve on this.

Separately, I'm wondering if reading 1 element per row is compatible with element-addressable memory work (I think it should be, but we should double-check).

### Cost estimates
I'd like to start thinking through some cost estimates for this. For example, how many hashes and memory reads would be required for a circuit of 5K gates and 200 inputs (I think these are ballpark numbers of what we'd need for the VM circuit). My initial estimates are:

- 5K gates require:
  - 5K rows in the hasher chiplet (8 gates can be "unhashed" in a single permutation).
  - 5K rows in the memory chiplet during evaluation (unless we can find a way to read 4 gates at a time).
- 200 inputs require:
  - 800 rows in the hasher chiplet (2 inputs can be "unhashed" in a single permutation).
  - 200 rows in the memory chiplet during input loading.

If we are verifying multiple proofs, circuit "unhashing" costs would be amortized. So, overall, verifying a proof would add between 6K and 11K rows to the chiplets table.

> It might be better to make the gates more expressive at the cost of adding more columns. The rational for this would be to decrease the gate count and hence the chiplet's execution trace length.

My intuition here is that we should keep the gates pretty simple: they are just additions/multiplications in the degree 2 extension field.

One potential exception to this could be something that could help us with MDS matrix multiplication because this part alone would contribute over 600 gates to the circuit.

**Comment by bobbinth** [comment]
> Another potential option is to encode all actions into `op`. This way, in addition to `add` and `mul` we'd also have `read` operation which would read an input from a given memory location.

This may actually work quite well. We could have 4 operations:

- `add` - addition in the extension field.
- `mul` - multiplication in the extension field.
- `read` - read the input from memory.
- `const` - set an input to some constant value.

One question is whether we also need a `sub` operation (it can be emulated with multiplying by $-1$, but maybe good to have a dedicated operation for it).

The layout could look something like this:
![image](https://github.com/user-attachments/assets/245ee4f5-1ec2-44ed-85df-0ef4b6bd8e77)

Where:
- The numbers under each element represent the number of columns - so, 15 columns total.
- I've assumed that we can arrange inputs and circuit to be next to each other in memory and so we'd need only 1 memory pointer. Not 100% sure if this will work though.

**Comment by Al-Kindi-0** [comment]
> Alternatively, we could skip m (the prover will need to fill it in in the chiplet) and then l_in_id, r_in_id, and out_id could be limited by 2^20.

I like this idea but it seems a bit counterintuitive to me that it could be sound though I can't immediately see a flaw in it.
This means that we can probably encode up to $2^{20}$ unique gate identifiers.

> Another potential option is to encode all actions into op. This way, in addition to add and mul we'd also have read operation which would read an input from a given memory location.

I think we should go with this option. As you describe in your second comment and if I understand correctly, this will require that the input data and circuit data be interleaved in memory. This is because for a `read` at address `addr` we would request a read at `addr + 1` and `addr + 2` in order to populate say `l_in_id` with the received value. Here I am assuming element addressable memory. Alternatively, we could just layout things so that things are word-aligned which might require some padding.
For a trace that is not too wide, insuring the above layout of input data and circuit data should not be too difficult. We can, after un-hashing the circuit description to memory, just populate the appropriate memory slots containing the circuit description with the input data. This should not be costly provided moderate trace widths.

> One thing that is not entirely clear to me is what's the best way to arrange inputs in memory. I believe currently we "interleave" current/next row values of the OOD point in a single word. Would this still work? Or would we need to make it so that each input value is stored in a separate word?

If we go with the solution in the previous bullet point then this should not be an issue as far as I can see.

> If I understood correctly, the described design assumes that we do 1 memory read of a single element per gate (i.e., per row of the chiplet). Since we can actually read 4 elements per memory read, I wonder if there is some way we can improve on this.

I think we should be able to read 1 element per gate once element addressable memory is done, but we will have to double check in the relevant PR.
Reading a word would be nice from a memory-chiplet-length point of view but I am not sure how to do it efficiently.

> 800 rows in the hasher chiplet (2 inputs can be "unhashed" in a single permutation).

When I think of input I think of a field element in the extension field, is it the same here?

> One potential exception to this could be something that could help us with MDS matrix multiplication because this part alone would contribute over 600 gates to the circuit.

We can resort to this if the circuits with only mul / add gates become too large.

**Comment by Al-Kindi-0** [comment]
> > Another potential option is to encode all actions into `op`. This way, in addition to `add` and `mul` we'd also have `read` operation which would read an input from a given memory location.
> 
> This may actually work quite well. We could have 4 operations:
> 
>     * `add` - addition in the extension field.
> 
>     * `mul` - multiplication in the extension field.
> 
>     * `read` - read the input from memory.
> 
>     * `const` - set an input to some constant value.
> 
> 
> One question is whether we also need a `sub` operation (it can be emulated with multiplying by − 1 , but maybe good to have a dedicated operation for it).
> 
> The layout could look something like this: ![image](https://private-user-images.githubusercontent.com/43513081/399191264-245ee4f5-1ec2-44ed-85df-0ef4b6bd8e77.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1NDM1NjcsIm5iZiI6MTczNTU0MzI2NywicGF0aCI6Ii80MzUxMzA4MS8zOTkxOTEyNjQtMjQ1ZWU0ZjUtMWVjMi00NGVkLTg1ZGYtMGVmNGI2YmQ4ZTc3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDA3MjEwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlM2UzM2VlNDE1OWM2MmMwZDFjYzZmOTZlMTI2ZDc5M2IyOTBiYmVkOWJjYTdjZTkzNmE0ZTBkYjg0ZWRhMGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Dof3Js3xaXokGLjm8c5XbyTDjdFKzsVC2siZzyqs95c)
> 
> Where:
> 
>     * The numbers under each element represent the number of columns - so, 15 columns total.
> 
>     * I've assumed that we can arrange inputs and circuit to be next to each other in memory and so we'd need only 1 memory pointer. Not 100% sure if this will work though.


One thing that is not clear to me from the above is what would the interface between the arithmetic evaluation chiplet and the decoder look like.

**Comment by bobbinth** [comment]
> > Alternatively, we could skip m (the prover will need to fill it in in the chiplet) and then l_in_id, r_in_id, and out_id could be limited by 2^20.
> 
> I like this idea but it seems a bit counterintuitive to me that it could be sound though I can't immediately see a flaw in it. This means that we can probably encode up to 2^20 unique gate identifiers.

Maybe including $m$ into the gate encoding is a safer way to go. I think 2^16 gates should be sufficient in most cases, and we can even push it to 2^18 if we reduce max value of $m$ to 1024.

> I think we should go with this option. As you describe in your second comment and if I understand correctly, this will require that the input data and circuit data be interleaved in memory. This is because for a `read` at address `addr` we would request a read at `addr + 1` and `addr + 2` in order to populate say `l_in_id` with the received value. Here I am assuming element addressable memory. Alternatively, we could just layout things so that things are word-aligned which might require some padding.
>
> For a trace that is not too wide, insuring the above layout of input data and circuit data should not be too difficult. We can, after un-hashing the circuit description to memory, just populate the appropriate memory slots containing the circuit description with the input data. This should not be costly provided moderate trace widths.

The way I was imagining this is that inputs would be aligned on word boundaries and each input word would be something like `[gate, 0, in_0, in_1]` where `gate` is the encoding of the gate and `(in_0, in_1)` would be the extension field element (this could also support degree 3 filed extension, if needed).

When computing the commitment to a circuit, input gates would be represented with full words `[gate, 0, 0, 0]`. This is a bit wasteful, but should simplify data movement.

Then, before invoking the chiplet, we'd need to populate the input values for each of the input cells - though, as you've mentioned, this may take quite a few cycles as we'd need to manually copy over each input into the appropriate gate.

The main benefit of this is that in the chiplet we need to do only 1 memory read per row:
- For `read` operation we'd read the entire word from memory and then we'd increment the memory pointer by 4.
- For `add`/`mul` operations we'd read a single element from memory and increment the memory pointer by 1.

At the same time, as mentioned above, copying over inputs will probably require quite a bit of data movement - so, maybe there is a better approach.

> > 800 rows in the hasher chiplet (2 inputs can be "unhashed" in a single permutation).
> 
> When I think of input I think of a field element in the extension field, is it the same here?

Correct. Here I was assuming we'd have an arrangement of single extension field element per word (with the "unused" elements set to zero). This is different from what we do now where we pack 2 field elements per word. With 2 elements per word we'd need only 400 rows, but I do think that there may be benefit to this field element arrangement (e.g., would be much easier to go to degree 3 extension).

> One thing that is not clear to me from the above is what would the interface between the arithmetic evaluation chiplet and the decoder look like.

I think the decoder would view it as a single instruction - something like:
```
#! Inputs: [circuit_ptr, circuit_end_ptr]
#! Output: [result_1, result_0]
eval
```
This instruction would need to be used in procedures which have some additional knowledge about the circuit being evaluated because these procedures would need to populate the inputs in memory. The process could look like so:

1. "Unhash" the circuit to memory (only needs to be done once per circuit).
2. "Unhash" the inputs to memory.
3. Populate the "input words" in the circuit description (but as mentioned above, maybe there is a better approach).
4. Call the `eval` instruction.

We'd have multiple procedures which look like this but for different circuits - e.g., `evaluate_vm_constraints`, `evaluate_keccak_constraints`, `evaluate_ecdsa_constraints`. Since each procedure is specialized, we should be able to use `repeat` loops rather than `while` loops which should make things a bit more efficient.

**Comment by Al-Kindi-0** [comment]
> Maybe including m into the gate encoding is a safer way to go. I think 2^16 gates should be sufficient in most cases, and we can even push it to 2^17 if we reduce max value of m to 1024.

Agreed! An upper bound of 1024 on m should be more than enough. This could be doubled inside the circuit at the cost of one additional gate if there is a need for a larger m sometime.

> When computing the commitment to a circuit, input gates would be represented with full words [gate, 0, 0, 0]. This is a bit wasteful, but should simplify data movement.

Yes, that's what padding was referring to. I think there are typically two extreme configurations that we face:

1. The VM trace which is narrow but has a complex constraint evaluation circuit. Here the number of inputs is manageable and hence both the amount of padding as well as the amount of data movement should be negligible when compared to the cost of evaluating the circuit.
2. Chiplet traces which are wide but usually have low complexity constraint evaluation circuits. Here the number of inputs will be large and hence we will incur a significant overhead due to the padding and data movement. On the other hand, the circuit will not be too complex.

All in all, I think we should be fine with the above proposal as it has the advantage of simplifying the design of the arithmetic circuit evaluation chiplet, namely reduction in the number of committed columns.

> Correct. Here I was assuming we'd have an arrangement of single extension field element per word (with the "unused" elements set to zero). This is different from what we do now where we pack 2 field elements per word. With 2 elements per word we'd need only 400 rows, but I do think that there may be benefit to this field element arrangement (e.g., would be much easier to go to degree 3 extension).

Makes sense!

> This instruction would need to be used in procedures which have some additional knowledge about the circuit being evaluated because these procedures would need to populate the inputs in memory. The process could look like so:

Makes sense! One more thing, how would the communication between the chiplet and the decoder work out? Requests from the decoder are straightforward but responses are not clear to me especially when we have multiple invocations of `eval` in the same execution trace of the chiplet.

**Comment by bobbinth** [comment]
> All in all, I think we should be fine with the above proposal as it has the advantage of simplifying the design of the arithmetic circuit evaluation chiplet, namely reduction in the number of committed columns.

If we remove $m$ from gate encoding, I think there is a good way for optimizing this even further. Here is how it could work:

Let's say we use `00` to encode `read` operation. Then, an input gate could be encoded like so:
```
0 | 00 (2 bits for op) | 40 zeros | out_id (20 bits)
```
This means that field element encoding value for wire with ID 0 would be 0, for ID 1 would be 1 etc.

Then, we lay out the circuit in memory like so:

![image](https://github.com/user-attachments/assets/46672a3d-3f4e-42b3-805b-95d2dd73f209)

In the input gate section (and probably in the other gate section as well), we make sure that `out` wire ID is always increasing (i.e., incremented by 1). So, the section could look something like this:
```
[0, 0, 0, 0], [1, 0, 0, 0], ..., [n, 0, 0, 0]
```
Where $0, 1, ..., n$ are the gate IDs.

Then, the procedure for evaluating the circuit becomes:

1. "Unhash" the circuit to memory (only needs to be done once per circuit).
2. "Unhash" the inputs to memory overwriting the "inputs" section of the circuit.
3. Loop through the inputs section and populate of the circuit and populate gate elements.
4. Call the eval instruction.

After step 2 above, the inputs section will look something like this:
```
[0, 0, v0_0, v0_1], [0, 0, v1_0, v1_1], ..., [0, 0, vn_0, vn_1]
```
Where $(v_0, v_1)$ are the degree 2 elements defining the inputs.

After step 3, we want the section to look like so:
```
[0, 0, v0_0, v0_1], [1, 0, v1_0, v1_1], ..., [n, 0, vn_0, vn_1]
```
Assuming the number of inputs is fixed (e.g., 200), we should be able to accomplish this with the following loop:
```
# => [input_ptr, gate_id = 0, ...]

repeat.200
    dup.1 dup.1 mem_store
    # => [input_ptr, gate_id, ...]

   add.4 swap add.1 swap
    # => [input_ptr + 4, gate_id + 1, ...]
end
```

This does take up ~10 cycles per input (so, for 200 inputs it'll be around 2K cycles, and for 2000 inputs, 20K cycles) - but I think that's not too bad.

If the above works, then the main thing to figure out is how to remove $m$ from gate definitions. I think this should be doable.

**Comment by Al-Kindi-0** [comment]
I am not sure the above would work as it assumes that we can un-hash the inputs to memory for free as 
```
[0, 0, v0_0, v0_1], [0, 0, v1_0, v1_1], ..., [0, 0, vn_0, vn_1]
```
The main reason is that, for our use case, the input will have other uses than just arithmetic circuit evaluation (ACE). For example, the OOD frame is used in the the computation of the random linear combination which goes into computing the queries to the DEEP polynomial.
This is a deal breaker currently with our use of `rcomb_*` but will not be so if we transition to the proposal of using `inner_prod_*`. Still, even if we transition to `inner_prod_*`, we are going to incur a cost for laying out OOD in a way that favors ACE as it will most certainly not be convenient for DEEP queries, i.e., not a deal breaker but with a cost.
To summarize, with the new proposal for `inner_prod_*` the above could be made to work but the savings in the cost for laying out the inputs as described above will imply, potentially, comparable costs in other places, e.g., DEEP queries, thus, potentially, negating the savings.

**Comment by bobbinth** [comment]
So, for OOD frame we need to do 2 things:

1. Provide OOD frame as an input to the arithmetic circuit evaluation chiplet.
2. Compute $\sum_{i=0}^k \alpha^{i + 1}\cdot T_i(z)$ and $\sum_{i=0}^k\alpha^{i + 1} \cdot T_i(z \cdot g)$ using the Horner evaluation instruction.

And the layout that would be suitable for circuit evaluation would make computing linear combinations difficult and vice versa. Unless we can come up with a way that works for both.

**Comment by Al-Kindi-0** [comment]
Indeed, and since bullet point 2 is more performance critical (as it is invoked number of queries many times) it should, I believe, always take precedence.

**Comment by bobbinth** [comment]
> bullet point 2 is more performance critical (as it is invoked number of queries many times)

I thought we need to do these only once as they only involve the OOD frame (and not the queries).

**Comment by Al-Kindi-0** [comment]
Indeed, that comment between parentheses is not correct.
So we would have to compare the costs of computing the sum in the second bullet point without the Horner evaluation instruction versus the cost of not using a layout the favor's  circuit evaluation.
There might, however, be a way to make Horner evaluation work with the circuit evaluation layout but will have to look more into it.

**Comment by Al-Kindi-0** [comment]
> There might, however, be a way to make Horner evaluation work with the circuit evaluation layout but will have to look more into it.

I think there is, will double check and write it soon.

**Comment by Al-Kindi-0** [comment]
Let $`(c_i)_{0\leq i\leq 2n-2}`$ be the coefficients of the polynomial represented with the proposed with-gaps layout in memory. This means that if $`(a_i)_{ 0 \leq i \leq n -1}`$ are the coefficients of the polynomial with the without-gaps layout then we have the following relations $`c_{2i + 1} = 0`$ and $`c_{2i} = a_i`$.

Now, we can use `horner_eval_ext` on the polynomial with coefficients $c_i$ to evaluate at some $x$ and this would give us $$c_0 + c_1 \cdot x + c_2 \cdot x^2 + \cdots + x^{2n-1} \cdot c_{2n - 1}$$

Using the above relation between $c_i$ and $a_i$ we get 
$$a_0 + a_1 \cdot x^2 + a_2 \cdot (x^2)^2 + \cdots + (x^2)^{n-1} \cdot a_{n - 1}$$

This means that evaluating $`(a_i)_{ 0 \leq i \leq n -1}`$ at $\alpha$ is equal to the evaluation of $(c_i)_{ 0 \leq i \leq 2n -2}$ at $\beta$ where $$\alpha = \beta^2$$

In other words, evaluating the polynomial with coefficients $(c_i)_{ 0 \leq i \leq 2n -2}$ at the square-root of $\alpha$ is equal to the evaluation of the original polynomial at $\alpha$.

Sadly, we are not always guaranteed to find $\beta$ for a given $\alpha$ but we can get an estimate on the probability of failing to do so.

**Comment by Al-Kindi-0** [comment]
Considering the cost of evaluating the x-independent term, assuming a trace of width 80 aux trace of width 8 and constraint trace of width 8, we can use roughly the following code:

```
# Since we have 80 main trace columns, 8 aux trace columns and 8 constraint columns, the number of terms
# in the sum related to `z` is 80 + 8 + 8 = 96. Since we process at each iteration 4 terms, the loop has to be 
# repeated 24 times.
repeat.24
    adv_pipe
    # => [0, 0, w1, w0, 0, 0, v1, v0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    movup.5 movup.5
    # => [0, 0, 0, 0, w1, w0, v1, v0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    swapw exec.constants::TMPX mem_storew
    # => [w1, w0, v1, v0, 0, 0, 0, 0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    adv_pipe
    # => [0, 0, w1`, w0`, 0, 0, v1`, v0`, C, data_ptr + 2, alpha_ptr, acc1, acc0, ...]

    movup.5 movup.5
    # => [0, 0, 0, 0, w1`, w0`, v1`, v0`, C, data_ptr + 2, alpha_ptr, acc1, acc0, ...]

    exec.constants::TMPX mem_loadw swapw
    # => [w1`, w0`, v1`, v0`, w1, w0, v1, v0, C, data_ptr + 2, alpha_ptr, acc1, acc0, ...]

    horner_eval_ext
end
```

```
# Since we have 80 main trace columns and 8 aux trace columns (note that there are no terms coming from
# the constraint trace in the sum associated to `zg`), the number of terms in the sum related to `zg` is 80 + 8 = 88. 
# Since we process at each iteration 4 terms, the loop has to be repeated 22 times.
repeat.22
    adv_pipe
    # => [0, 0, w1, w0, 0, 0, v1, v0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    movup.5 movup.5
    # => [0, 0, 0, 0, w1, w0, v1, v0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    swapw exec.constants::TMPX mem_storew
    # => [w1, w0, v1, v0, 0, 0, 0, 0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    adv_pipe
    # => [0, 0, w1`, w0`, 0, 0, v1`, v0`, C, data_ptr + 2, alpha_ptr, acc1, acc0, ...]

    movup.5 movup.5
    # => [0, 0, 0, 0, w1`, w0`, v1`, v0`, C, data_ptr + 2, alpha_ptr, acc1, acc0, ...]

    exec.constants::TMPX mem_loadw swapw
    # => [w1`, w0`, v1`, v0`, w1, w0, v1, v0, C, data_ptr + 2, alpha_ptr, acc1, acc0, ...]

    horner_eval_ext
end
```

Note that the cost for the 
```
[0, 0, v0_0, v0_1], [0, 0, v1_0, v1_1], ..., [0, 0, vn_0, vn_1]
```
alignment is $15$ cycles per iteration. Since we have $24 + 22 = 46$ iterations, the total overhead can be estimated as $46 * 15 = 690$ cycles.

**Comment by plafer** [comment]
> Maybe including m into the gate encoding is a safer way to go. I think 2^16 gates should be sufficient in most cases, and we can even push it to 2^17 if we reduce max value of m to 1024.

> If we remove m from gate encoding, I think there is a good way for optimizing this even further. Here is how it could work:

I think it's safe to remove `m` from the gate encoding as I don't see how to prover could cheat. For a given wire, `m` is equivalent to the number of gates that use that wire as input. And the prover is forced to read all the gates from memory. So if the prover tries to cheat and use an incorrect value for `m`, then the bus will simply fail (since the prover is forced to populate it properly with all the gates that use the wire as input). So it's probably fine to do as initially proposed and use `2^20` as an upper bound for the 3 values (without encoding `m`)?

**Comment by Al-Kindi-0** [comment]
Yeah, for a committed sound circuit, there should be one and only way for the prover to fill in the `m` column.
So we can thus go to a maximum of $2^{20}$ gates per evaluated circuit.

**Comment by bobbinth** [comment]
Yes, I think we can safely get rid of $m$ from gate descriptions. The way I imagine this will work in the processor is as follows:

- When the processor sees the `eval_circuit` instruction, it reads the circuit from memory and builds an internal representation of this circuit. In this internal representation, it will track how many times a given wire has appeared as an input into a gate. This will give the processor $m$ values for all wires.
- The internal representation of the circuit can then be used to evaluate the circuit itself quickly and put the results onto the stack.
- It can also be passed to the chiplet trace builder to correctly insert the value of $m$ for every gate.

Regarding computing $\sum_{i=0}^k \alpha^{i + 1}\cdot T_i(z)$ and $\sum_{i=0}^k\alpha^{i + 1} \cdot T_i(z \cdot g)$ - don't we need to compute theme separately? Meaning:

- First, we iterate over values of $T_i(z)$ to compute $\sum_{i=0}^k \alpha^{i + 1}\cdot T_i(z)$.
- Then, we iterate over values of $T_i(z \cdot g)$ to compute $\sum_{i=0}^k\alpha^{i + 1} \cdot T_i(z \cdot g)$.

So, our current "zig-zag" arrangement of $T_i(z)$ and $T_i(z \cdot g)$ wouldn't work. Instead, we'd need to lay out the OOD frame to first have all the $T_i(z)$ values and then $T_i(z \cdot g)$ values.

I think the approaches outlines in https://github.com/0xPolygonMiden/miden-vm/issues/1610#issuecomment-2581126473 could work. Another potential approach is to have `horner_eval_ext` instruction work only with 2 elements at a time (instead of 4). So, it could look something like this (assuming 88 values):

```
# compute sume of alpha^{i + 1} * T_i(z)
repeat.44
     adv_pipe
    # => [0, 0, w1, w0, 0, 0, v1, v0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    hperm
    # => [0, 0, w1, w0, 0, 0, v1, v0, C', data_ptr, alpha_ptr, acc1, acc0, ...]

    # compute acc' = ((acc + v) * alpha + w) * alpha - please double check this formula
    horner_eval_ext
    # => [0, 0, w1, w0, 0, 0, v1, v0, C, data_ptr + 8, alpha_ptr, acc1', acc0', ...]
end

# TODO: save acc to memory

# compute sum of alpha^{i + 1} * T_i(z * g)
repeat.44
     adv_pipe
    # => [0, 0, w1, w0, 0, 0, v1, v0, C, data_ptr, alpha_ptr, acc1, acc0, ...]

    hperm
    # => [0, 0, w1, w0, 0, 0, v1, v0, C', data_ptr, alpha_ptr, acc1, acc0, ...]

    # compute acc' = ((acc + v) * alpha + w) * alpha - please double check this formula
    horner_eval_ext
    # => [0, 0, w1, w0, 0, 0, v1, v0, C, data_ptr + 8, alpha_ptr, acc1', acc0', ...]
end

# TODO: save acc to memory
# TODO: verify that OOD frame was un-hashed correctly
```

Even though we now need to do 44 iterations (and 88 total for the OOD frame), each iteration iteration requires only 3 cycles. And so, we get a total of 264 cycles to both unhash the OOD frame and to compute  $\sum_{i=0}^k \alpha^{i + 1}\cdot T_i(z)$ and $\sum_{i=0}^k\alpha^{i + 1} \cdot T_i(z \cdot g)$.

There are probably some details I'm missing, but accounting for these, I think we should be able to perform this part of the process in under 500 cycles.

**Comment by Al-Kindi-0** [comment]
Here is an attempt at a summary of the discussion so far. There is also an addition related to the ops of the ACE chiplet, but this is more of an initial proposal/suggestion rathar than a final decision.

# Layout of the chiplet
1. Opcode bits: $s_0, s_1, s_2 \in \{0, 1\}$ 
2. Input gate ids: `l_in_id, r_in_id, out_id` are integer values in the range $\{0, 1, \cdots, 2^{20} - 1\}$
3. Input gate values: `l_in_val, r_in_val, out_val` are each elements in the quadratic extension field.
4. Gate output arity: `m` which indicates the number of output wires emanating from the current gate.
5. Memory pointer: `mem_ptr` are integer values in the range $\{0, 1, \cdots, 2^{32} - 1\}$.
6. Clock cycle: `clk` are integer values in the range $\{0, 1, \cdots, 2^{32} - 1\}$.
7. Memory context: `ctx` are integer values in the range $\{0, 1, \cdots, 2^{32} - 1\}$.

There is also a an auxiliary column acting as a bus and used for wiring the gates of the circuit. The equation defining this bus is
$$b^{'} = b +  \frac{m}{\alpha -(id_{\textsf{out}} + \beta\cdot v_{\textsf{out}})} - \frac{1}{\alpha -(id_{\textsf{l-in}} + \beta\cdot v_{\textsf{l-in}})} - \frac{1}{\alpha -(id_{\textsf{r-in}} + \beta\cdot v_{\textsf{r-in}})}$$

where the selectors in the numerators are omitted for now.

# Opcodes

We will describe the (internal) opcodes of the ACE chiplet. The big picture of the design can be summarized in the following points:
1. The evaluation of an arithmetic circuit involves 3 stages, reading inputs, evaluating the circuit and outputing the result. There is also a step which should indicate the start of the evaluation of a specific circuit, call it the initialization stage.
2. The memory pointers are managed by the chiplet and hence we need to enforce the correctness of this using constraints. This boils down to three points:
   1. For the reading inputs stage, the pointers have to be incremented in increments of 4.
   2. For the evaluation stage, the pointers have to be incremented in increments of 1 except for when the gate is a gate loading a constant, where we increment the pointer by 2 to account for the immediate value i.e., the constant in the base field.
   3. We need to make sure that reading inputs is done before evaluating the circuit and that both stages are contiguous.
3. We need to wire the gates in order to feed the outputs of each gate to the gates to which they are inputs. For this we use the bus mentioned above. Note that the `m` cell values are filled by the prover.
## INIT
This represents the initialization stage which is the beginning of a new arithmetic circuit evaluation cycle.

1. Send a response to the decoder indicating the beginning of a new evaluation cycle, the response will have the following shape `(START_EVAL_LABEL, mem_ptr, clk, ctx)`
2. Next op should be a `READ`
3. `mem_ptr' = mem_ptr`
4. `clk' = clk`
5. `ctx' = ctx`
6. `out_id = 0`
7. `out_id' = out_id`

## READ
This represents the reading of the input gates values which are represented each as a word.

1. Make a memory request to read a word at address `(ctx, mem_ptr)` at cycle `clk`. The word to be read is `[out_id, 0, out_val0, out_val1]`. Note that we will define the opcode as `READ = (0, 0, 0)` and thus the first element of the word, which is computed as $s_2 \cdot 2^{62} + s_1 \cdot 2^{61} + s_0 \cdot 2^{60} + v_2 \cdot 2^{40 } + v_1 \cdot 2^{20} + v_0$ where `(v_0, v_1, v_2) = (out_id, l_in_id, r_in_id)`, becomes just `out_id`.
2. Next op should either be a `READ` or an `EVAL`
3. `mem_ptr' = mem_ptr + 4`
4. `clk' = clk`
5. `ctx' = ctx`
6. `out_id' = out_id + 1`

We also need to wire these input gates through the wiring bus and we do this by enforcing the following equation for all input gates
$$b^{'} = b +  \frac{m}{\alpha -(id_{\textsf{out}} + \beta\cdot v_{\textsf{out}})}$$

## EVAL

This represents any of the four ops `MUL, ADD, SUB, CONST`. For all these ops we have to enforce that either the next op is an `EVAL` or is an `OUTPUT`.
### MUL, ADD, SUB
1. Make a memory read to read an element at address `(ctx, mem_ptr)` at cycle `clk`. The element to be read is $s_2 \cdot 2^{62} + s_1 \cdot 2^{61} + s_0 \cdot 2^{60} + v_2 \cdot 2^{40 } + v_1 \cdot 2^{20} + v_0$ where `(v_0, v_1, v_2) = (out_id, l_in_id, r_in_id)` 
2. Assert the gate algebraic relation implied by $(s_0, s_1, s_2)$ between `l_in_val, r_in_val` and `out_val`.
3. `mem_ptr' = mem_ptr + 1`
4. `clk' = clk`
5. `ctx' = ctx`
6. `out_id' = out_id + 1`

The bus equation for such gates is
$$b^{'} = b +  \frac{m}{\alpha -(id_{\textsf{out}} + \beta\cdot v_{\textsf{out}})} - \frac{1}{\alpha -(id_{\textsf{l-in}} + \beta\cdot v_{\textsf{l-in}})} - \frac{1}{\alpha -(id_{\textsf{r-in}} + \beta\cdot v_{\textsf{r-in}})}$$

### CONST
Same as previous point but here we want to use an immediate value and hence we will have to make an additional element-read request to memory at address `(ctx, mem_ptr)` at cycle `clk` for the element in cell `out_val`. This means that we now have to enforce `mem_ptr' = mem_ptr + 2`.

The bus update constraint in this case will look like
$$b^{'} = b +  \frac{m}{\alpha -(id_{\textsf{out}} + \beta\cdot v_{\textsf{out}})}$$

## OUTPUT
This will be selected for by `OP' = OUTPUT`, and here we mainly send a response to the decoder indicating the conclusion of an evaluation. The response can have the following shape `(END_EVAL_LABEL, mem_ptr, clk, ctx)` (note that we are looking at next op i.e., `OP'`).

Note that we will have to balance the wiring bus somehow by either subtracting $\frac{1}{\alpha -(id_{\textsf{out}} + \beta\cdot v_{\textsf{out}})}$ when `OP' = OUTPUT` or by multipliying $\frac{m}{\alpha -(id_{\textsf{out}} + \beta\cdot v_{\textsf{out}})}$ in the above expressions by a selector that is $1$ except for when `OP' = OUTPUT`

# Usage

## Pre-processing stage
This stage is done by the verifier or a trusted party once and only once. The verifier will build the circuit it wishes to evaluate in the format described above, namely it will fill-out the op-code columns $s_0, s_1, s_2$, the input gate ids columns `l_in_id, r_in_id, out_id` thus generating a matrix where each row represents a gate of the aforementioned circuit.
The verifier will then encode each gate using the encoding discussed above i.e., for gates with `OP = READ`, the gate will be represented as a word of the form `[out_id, 0, 0, 0]` where `out_id` is in the range $\{0, 1, \cdots, 2^{20} - 1\}$, for gates with `OP = ADD or MUL or SUB` the gate will be represented by an element $s_2 \cdot 2^{62} + s_1 \cdot 2^{61} + s_0 \cdot 2^{60} + v_2 \cdot 2^{40 } + v_1 \cdot 2^{20} + v_0$ where `(v_0, v_1, v_2) = (out_id, l_in_id, r_in_id)`, and for gates with `OP = CONST` the gate will be represented by two elements where the first one is as in the case of the other `EVAL` ops and the second one will represent the constant value.
During the building of the circuit, the verifier will learn the maximal value that the column `mem_ptr` will take i.e., the last value just before `OUTPUT` as the values in this column will be monotonically increasing during the course of a computation cycle. From this upper bound on `mem_ptr`, the verifier will extract a constant, call it `N`, that will, together with the hash of the circuit just computed, identify the arithmetic circuit in question.

## In the VM
Suppose we want to evaluate an arithmetic circuit `(N, DIGEST)`, then the process could look like:
1. Unhash the circuit to some word-aligned region of memory.
2. Unhash the inputs to the `READ` section (i.e., the section where `OP` is `READ`) of the previous memory region thus overwriting it with the inputs. The `READ` section will then look like `[0, 0, v0_0, v0_1], [0, 0, v1_0, v1_1], ..., [0, 0, vn_0, vn_1]`
3. Loop through the `READ` section and set the first element to the correct `out_id` value. After this step, the `READ` section will look like `[0, 0, v0_0, v0_1], [1, 0, v1_0, v1_1], ..., [n, 0, vn_0, vn_1]`. This effectively sets the op of this gate to `READ`. See section on `READ` op.
4. Call the eval instruction with stack inputs `circuit_start_ptr` and `N`, where `circuit_start_ptr` is the start of the memory region storing the unhashing of the circuit with digest `DIGEST` and `N` is the memory adress upper bound defining the said circuit. The eval instruction will then:
   1. Define `circuit_end_ptr := circuit_start_ptr + N`
   2. Make a double request to the circuit evaluation circuit, the first is to initiate `(START_EVAL_LABEL, circuit_start_ptr, clk, ctx)` and the second is to request the result `(END_EVAL_LABEL, circuit_end_ptr, clk, ctx)`.


# TODOs

The number and encoding of the opcodes is still to be determined/finalized. One thing to consider is to maybe do away with `SUB` as we can negate by multipliying by `-1` in the field. This way we can give all the `EVAL` ops opcodes with prefix `0`.

**Comment by Al-Kindi-0** [comment]
> I’m not entirely sure I get how `CONST` would work since it requires reading two elements. Would we read two elements in a single row or spread the instruction over two consecutive ones? For uniformity, it almost seems that two rows is easier, as that would be symetric with the circuit description, where a const instruction is encoded as the two elements `[OP_CONST, const_val]`. The `CONST` opcode would be the only one in which we do not increment `out_id`, though it may actually be desirable so that `out_id` at the end represents the number of gates in the chip. The bus update for `out_val` would take `m=0`.

I was thinking we read two elements in one row. This will make the interactions with the memory a bit non-uniform but I think that the degrees work fine nevertheless.


> Regarding the program portion that prepares the memory, I wonder if we could avoid encoding the `READ` instruction and replace them the following. We describe the circuit with an additional variable `M` representing the number of inputs. This should be equal to `out_id-1` in the first `EVAL` operation, so we may be able to avoid providing directly. Before calling the chiplet, we would start by preparing the `READ` instructions by storing in memory the words `[i, 0, vi_0, v1_1]` for `i = 0..M`, corresponding the `M` `READ` instructions, starting at `circuit_start_ptr`. The circuit is unhashed into the address space following the `READ`s, i.e. `circuit_start_ptr+M`, and then perform step 4 of “In the VM” as before, with the same value of `circuit_end_ptr`.

If I understand correctly, the idea is to not have a `READ` opcode at all, right? In this case I am not sure how things would work in the chiplet with this modification. Probably the idea is to use the `M` you mentioned but would be good to get more clarity on this.

> It’s unclear to me still though whether we could take advantage of the above to interleave the preparation of the `READ`s with the Horner evaluation of the OOD frames.

I think we should be able to do this. Basically as we load OOD frames from the advice provider, we:

1. Compute the inner product with the powers of the batching randomness using `horner_eval_ext`
2. We update the gate ids. This would probably require the use of memory as the operand stack would be full from point 1
3. Store in memory using `mem_stream`

Note that the above will require that we modify `horner_eval_ext` in order to work with the layout  `[0, 0, vi_0, v1_1]`, but this is something we discussed in another issue and it seems it would be fine.

> The idea of doing away with `SUB` is interesting as it might remove a selector, though I feel that it might actually result in many more rows, as negation is very common in constraints (any `assert_eq`). An alternative would be to have a single `MULADD` instruction of arity 3 computing `out = a + b * c`. The issue is that we now perform 4 bus operations, and we need to encode another id inside the instruction. However, it feels to me that we would be able to avoid including the `out_id` since we can assume it’s always increasing. Need to think more about this though.

Actually in many situations, the expressions can be rearranged in order to avoid or at least minimize `SUB`. For example:

1. $X \cdot (1 - X) = X \cdot (X - 1) = X \cdot (X + \gamma)$ where $\gamma = -1$. This basically holds for any term of the form $X - k$ e.g., the constraints on the range checks column.
2. Flags have terms of the form $1 - f$ and these can be written as $f - 1 = f + \gamma$
3. All randomized constraints with $\alpha - v$ can be written as $\alpha + v$
4. ...

Basically, I think the main places where we would have to add `SUB` emulation rows in the chiplet is when we have terms `X - Y` for variables `X` and `Y`.

**Comment by adr1anh** [comment]
> I was thinking we read two elements in one row. This will make the interactions with the memory a bit non-uniform but I think that the degrees work fine nevertheless.

That's fair, and likely easier to implement in the first place. Is it correct to assume that a chiplet will always have exclusive access to 1 aux column? If so then having an additional memory interaction should be fine given the large degree we can work with. 

I think we can then assume that `out_id` increases by 1 in every row, though we still need to account for the fact that the first and last row need to perform decoder requests to indicate the start and end of the circuit evaluation. 

I'm still wondering therefore if we need to include `out_id` in the `OP`, as it essentially just encodes the length of the circuit (number of inputs and operations). 

> If I understand correctly, the idea is to not have a READ opcode at all, right? In this case I am not sure how things would work in the chiplet with this modification. Probably the idea is to use the M you mentioned but would be good to get more clarity on this.

The chiplet trace still includes the `READ` rows, since it needs to be able read the inputs from memory and send them to the circuit bus. My suggestion was that we don't need to encode these in the hash of the circuit. We just need to know the number of inputs and operations. Moreover I think we can also group all constants into a region between the inputs and the operations, and load them into the circuit bus in a similar way to the inputs. I'll write in the next comment a more complete description of what I have in mind.

> I think we should be able to do this. Basically as we load OOD frames from the advice provider, we:
> [...]

That sounds good, though I think if my next post works out, that we'd be able to store two input points in a single word, which would make it a bit easier to use the existing Horner operation.

> Actually in many situations, the expressions can be rearranged in order to avoid or at least minimize `SUB`. For example:
> [...]

That makes sense, and this is likely something that we can optimize for at the AirScript level to ensure we limit the number of negations.

**Comment by adr1anh** [comment]
The main idea is that the circuit can be split into three phases
1. Reading inputs given as the OOD evaluations
2. Reading constants embedded in the circuit
3. Evaluating the operation describing the circuit

%% I'm going to assume - though I'd like to make sure my intuition is correct - that we don't need to embed the `out_id` in the circuit, since we can enforce that this value increases by a fixed amount one with every read or evaluation.  %%

The number of inputs and constants are properties of the circuit which can be provided along with, or as part of, the digest. Let's call them `I` and `C` and assume they are even. We define `G` as the number of arithmetic operations.

The procedure starts by loading all inputs as extension field elements in the memory at address `mem_ptr`, with two extension field elements per word. This makes it more efficient to use the existing Horner operation implementation. The memory pointer increases by `I/2` (we can always append a garbage input for alignment, which would never be read by the circuit).

We then start unhashing the digest, which we know starts with `C` constants. While not necessarily useful, we can consider constants in the extension field rather than the base field. This ensures the region of memory containing the constants has the same alignment as the inputs, filling it with zeros if necessary. The memory pointer increases by `C/2`.

After the constants, the remainder of the digest containing all operations is unhashed, such that each word in memory contains 4 operations, embedding the opcode `ADD` or `MUL`, along with `(ID_L, ID_R, ID_O)` which are all $<2^{20}$. The memory pointer increases by `G/4`, which we can ensure is a multiple of 4 by appending additions of zero to the end result.

To recap, the memory will look like the following, where the total number of rows is `I/2 + C/2 + G/4`. 

|              | `addr` | `addr + 1` | `addr +2` | `addr + 3` |
| ------------ | ------ | ---------- | --------- | ---------- |
| `inputs_ptr` | `u[0]` | `u[1]`     | `v[0]`    | `v[1]`     |
| `consts_ptr` | `u[0]` | `u[1]`     | `v[0]`    | `v[1]`     |
| `ops_ptr`    | `op_0` | `op_1`     | `op_2`    | `op_3`     |


## Chiplet description


The layout of the chiplet can be viewed as two separate blocks, since we will treat inputs and constants in the same way. The layout of the columns would look something like this:

| 0         | 1     | 2     | 3     | 4               | 6       | 7      | 5        | 8     | 9            | 10           | 11         | 12         | 13         | 14         |
| --------- | ----- | ----- | ----- | --------------- | ------- | ------ | -------- | ----- | ------------ | ------------ | ---------- | ---------- | ---------- | ---------- |
| `is_read` | `ptr` | `clk` | `ctx` | `n_inputs`      | `n_inv` |        | `id_out` | `m_u` | `u[0]`       | `u[1]`       |            | `m_v`      | `v[0]`     | `v[1]`     |
| `is_eval` | `ptr` | `clk` | `ctx` | `is_{add, mul}` | `id_l`  | `id_r` | `id_out` | `m`   | `val_out[0]` | `val_out[1]` | `val_l[0]` | `val_l[1]` | `val_r[0]` | `val_r[1]` |

The circuit is considered evaluated and correct by checking the initial and final `ptr` value, which confirms that the entire memory has been consumed and the multi-set containing the variables evens out. Moreover, we ensure that `clk` and `ctx` remain the same in every row.

What enables us to pack two values per row during the read phase is that we can avoid encoding the `READ` instruction in the memory. Instead, the circuit can be initialized with the number of `READ` instructions it must perform before switching to the `is_eval` phase where we read the memory as arithmetic operations. 

We give an overview of the constraints for both phases, dictated by the flag `is_read/is_eval` which are the two cases of the selector. Note the these constraints have not been fully checked.

#### `is_read`

In the first row of the circuit, the following must happen 
- Respond to the `(BEGIN_EVAL_OP_LABEL, clk, ctx, ptr, n_inputs)` message that requested the circuit evaluation
- Assert `is_read = 1`,  ensuring we start with the `READ` operations
- Assert `id_out = 0`, ensuring the index of the leaf nodes start from 0.

For each row where `is_read = 1`
- Read the word `[u[0], u[1], v[0], v[1]]` from memory at address `(ctx, ptr)` at cycle `clk`.
	- Increment the pointer by the word size `ptr' = ptr + 4`
- Send to the ACE bus the entries 
	- `(id_out, v[0], v[1])` with multiplicity `m_v`
	- `(id_out + 1, u[0], u[1])` with multiplicity `m_u`
	- Increment the next variable index `id_out' = id_out + 2`.
- When the next row is a read, decrement the read instruction counter `n_input' = n_input - 1`.
- Ensure that we transition to the `is_eval` block if and only if `n_input` has reached zero
	- Assert `n_input * n_inv - is_read' = 0`
		- When the next row is a read, `n_input` must be non-zero with inverse `n_inv`.
	- Assert `is_eval' * n_input = 0`
		- When we transition to the `is_eval` block, `n_input` must have reached zero.
		- If `n_input` has not reached zero, the next row must be a read, which will decrement the counter

#### `is_eval`

This portion is mostly the same as what @Al-Kindi-1 has written up previously. In each row where `is_eval = 1`

- Read the element `op = pack(is_{add,mul}, id_l, id_r, id_out)` from memory at address `(ctx, ptr)` at cycle `clk`.
	- Increment the pointer by the element size `ptr' = ptr + 1`.
- In the ACE bus
	- Pull the left value `(id_l, val_l[0], val_l[1])` 
	- Pull the right value `(id_r, val_r[0], val_r[1])` 
	- Push the output value `(id_out, val_out[0], val_out[1])` with multiplicity `m`
	- We don't need to increment the `id_out`, but if we do, we can avoid including it in `op`.
- Check the correctness of the arithmetic operation
- Ensure that the next row is not a read

In the last row, we need to 
- Push the message `(END_EVAL_OP_LABEL, clk, ctx, ptr)` indicating the end of the circuit evaluation
- Perform some terminal assertion on `v_out=0`, or output this value as part of the message to let the caller check its correctness

## Closing observations

While this approach to reading inputs allows for more efficient utilization of memory, it comes at a tradeoff of more bus interactions and needing to supply inverses (though these are very easy to batch as they are sequential). Making sure that all these interactions don't increase the logUp constraint degree must still be checked.

It still seems that including the `id_out` in the operation is unnecessary. If it was used to refer to an existing node, thereby creating a back-link of some sort in the evaluation DAG, then we could just set `m=0` to disregard the assertion. Removing it would increase the total number of gates, though it's not clear whether $2^{30}$ is really necessary. 

The other advantage is that we reduce the size of the circuit, since constants only require 1 field element instead of 2. This reduces the number of hashes required to represent the circuit. 

We should also be able to avoid encoding a `CIRCUIT_{START, END}` flag in the `op` flag, as it could just be a selector in the trace. This flag would only be used to indicate if we're at the boundary of the trace for a specific circuit, and would activate the bus for sending the `START/END` message back to the CPU. If these are set incorrectly, then the wrong bus message is sent because the `ptr` would be wrong.

**Comment by Al-Kindi-0** [comment]
> The number of inputs and constants are properties of the circuit which can be provided along with, or as part of, the digest. Let's call them `I` and `C` and assume they are even. We define `G` as the number of arithmetic operations.

Got it! This is basically the same idea with regards to the input `N` defining the circuit together with the circuit `DIGEST`. With your proposal, we would add in addition to `N` and `DIGEST` `I + C` and `G`, probably `N` becomes redundant now so we can remove it as we can get it from `I + C` and `G`


> While this approach to reading inputs allows for more efficient utilization of memory, it comes at a tradeoff of more bus interactions and needing to supply inverses (though these are very easy to batch as they are sequential). Making sure that all these interactions don't increase the logUp constraint degree must still be checked.

I think this is the big question we have to answer. With this proposal, we are making really good use of the "real estate" on the chiplet but at the cost of making some columns multi-purpose and the constraints non-uniform which will for sure increase the degrees. Nevertheless, we should be able to overcome this even if at the cost of adding degree reducing columns. This is because being able to store two extension field elements per word for the inputs will allow us to have a 2x the throughput in `horner_eval_ext` than the original proposal. This might not be a big deal in narrow traces but will become a non-negligible chunk for wide ones.

> It still seems that including the `id_out` in the operation is unnecessary. If it was used to refer to an existing node, thereby creating a back-link of some sort in the evaluation DAG, then we could just set `m=0` to disregard the assertion. Removing it would increase the total number of gates, though it's not clear whether 2 30 is really necessary.

I am leaning towards the same conclusion. As long as the pointers are constrained to start and increase correctly and the `id_out` is constrained to increase correctly then I can't see how one can cheat i.e., it is not feasible to read things out of order or from a different place altogether.

> The other advantage is that we reduce the size of the circuit, since constants only require 1 field element instead of 2. This reduces the number of hashes required to represent the circuit.

Agreed!

> We should also be able to avoid encoding a `CIRCUIT_{START, END}` flag in the `op` flag, as it could just be a selector in the trace. This flag would only be used to indicate if we're at the boundary of the trace for a specific circuit, and would activate the bus for sending the `START/END` message back to the CPU. If these are set incorrectly, then the wrong bus message is sent because the `ptr` would be wrong.

That was my original thinking but I think I switched to the current proposal because I wasn't sure if the above would work when we are evaluating multiple circuits per (main) execution trace.

**Comment by adr1anh** [comment]
Summarizing a discussion with @Al-Kindi-0 from earlier.

We agree on the proposal described in the [previous reply](https://github.com/0xPolygonMiden/miden-vm/issues/1610#issuecomment-2722011864), as long as we can make sure it respects the degree bound. Most constraints seem fine, but we need to double-check that the extra wiring interactions with logUp can fit in a single aux columns. If I understood correctly, the chiplet and memory interactions affect a separate aux column which uses a running product. 

We agree that by using unique and strictly increasing `id_out` values is sound, so it can be removed from the `EVAL` instruction packing. Its value is constrained by the memory address constraints and the layout of the instructions (`READ`s followed by `EVAL`s). This technically allows for labels up to $2^{30}$ though it's unlikely we would encounter these in practical situations. 

## Boundary constraints for multiple ACE executions 

We want to be able to evaluate multiple constraints for different circuit within the same chip though we likely need one or more extra selector columns to define the start and end row of each evaluation.

We might be able to use only a single one, defined as `is_boundary` which is set to 1 in the first and last row of the chip. From it, we derive degree-2 selectors
- `is_start = is_boundary * is_read`
- `is_end = is_boundary * is_eval`
These are only used for the chiplet communication bus, i.e. 
`b_chip' * [ (1-is_start) + is_start * v_start ] = b_chip * [ (1-is_end) + is_end * v_end`, where `v_{start, end}` are the contents of the start and end messages. Note that here, `is_read` and `is_eval` are derived from the same selector defining the operation being performed. The `is_boundary` selector must be correct, since it would otherwise cause a mismatch between the main trace and chiplet messages. We would also use it to initialize the chip, requiring 
- `is_start * (1 - is_read) = 0` to ensure we start the chiplet with the `READ` operations.
- `is_start * id_out = 0` to ensure the `out` variables start with 0.

There is one issue that we need to tackle still. A prover to cheat by including an arbitrary number of `READ` rows before responding to the `START` message. This will load fake inputs at `ptr-4` and `ptr-2` at indices `-2` and `- 1` respectively, or going back any arbitrary number of rows in memory. This means the indices `id_l, id_r` could also be manipulated to be different than those encoded in the operation. (Similarly, we could continue the `EVAL` by reading beyond the final row, though this does not seem to cause soundness issues).

It seems enough to add the condition `is_read * (1 - is_start') = 0`, to ensure that we can never start the circuit after starting the `READ` block. For extra assurance, we can make it symmetrical with the `EVAL` block by requiring `is_end * (1 - is_eval') = 0` so that the row after ending the computation cannot continue evaluating. 

## Separating the wiring buses

If we allow multiple ACE executions, we need to make sure the wiring does not interfere across evaluations. We will need to add `(ctx, clk)` along with the index and value of each wire.

## Subtraction without additional selector

Given the extra space in the instruction value, we could recover the ability to perform subtractions by encoding the operation as `{-1, 0, 1}`, where 
- `-1` is subtraction
- `0` is multiplication
- `1` is addition

This leads to a simplified arithmetic operation of the form 
`v_out = op * [ val_l + op * val_r] - (op - 1) * (op + 1) * [ val_l * val_r ]`.
The degree is 4, so it should fit within the degree bound even when we multiply it with extra selectors.

## Nice to have

If we want to evaluate the same circuit multiple times, we only need to un-hash the digest once. However we still need to reread the memory for the constants and operations and copy it in the region after the different set of inputs. While this could save quite a few cycles, it's not clear at the moment what the best way to achieve this would be. This is not crucial for finalizing the design so we will leave it for a future optimization.

One approach Al mentioned was that we could send the evaluation directly from the stack to the ACE chiplet after hashing them into the transcript and accumulating them with the Horner operation.

**Comment by adr1anh** [comment]
The first draft of the full chiplet design can be found [here](https://hackmd.io/@adr1anh/BJH4n_KnJg). It formalizes the design proposed earlier and describes all the constraints and columns required. 

TL;DR: The chiplet requires 16 internal columns, and has a maximum internal degree of 5, which means it could be placed in 3rd or 4th position in the trace.

**Comment by Al-Kindi-0** [comment]
> The first draft of the full chiplet design can be found [here](https://hackmd.io/@adr1anh/BJH4n_KnJg). It formalizes the design proposed earlier and describes all the constraints and columns required.
> 
> TL;DR: The chiplet requires 16 internal columns, and has a maximum internal degree of 5, which means it could be placed in 3rd or 4th position in the trace.

Looks great!
Left just a few remarks but otherwise I don't see any significant issues.

One thing that I have just revisited is the question of transforming the problem to checking that a circuit evaluates to $0$ instead of any value. This would allow the whole constraint check to be offloaded to the ACE chiplet and hence reduce further the stack/decoder cycle count for the recursive verifier.

The way to do this is simple, we just remove all the denominators. Basically, we write 

$$\frac{\mathsf{num}_0 }{\frac{z^n - 1}{z - g^{-1}}} + \frac{\mathsf{num}_1}{z-1} + \frac{\mathsf{num}_2}{z - g^{-1}} = Q(z)$$

as

$$ (z -  g^{-1})^2 (z - 1) \mathsf{num}_0 + (z^n - 1) \left( (z - g^{-1}) \mathsf{num}_1 + (z - 1) \mathsf{num}_2 \right)  - Q(z)  (z^n - 1)  (z - 1)  (z -  g^{-1}) = 0$$

**Comment by Al-Kindi-0** [comment]
Looks good to me and I am in favor of including this change in the draft.

**Comment by Al-Kindi-0** [comment]
@adr1anh , I left just one minor comment in the updated draft, otherwise I think we can close this issue as completed as any further discussions can be done in the follow up issues/PRs, what do you think?
